haas/api.py

```diff
@@ -258,34 +275,57 @@ def node_connect_network(node, nic, network):
 
     project = node.project
 
+    allocator = get_network_allocator()
+    if channel is None:
+        channel = allocator.get_default_channel(db)
+
     if nic.current_action:
         raise BlockedError("A networking operation is already active on the nic.")
 
     if (network.access is not None) and (network.access is not project):
         raise ProjectMismatchError("Project does not have access to given network.")
 
-    db.add(model.NetworkingAction(nic, network))
+    if _have_attachment(db, nic, model.NetworkAttachment.network == network):
+        raise BlockedError("The network is already attached to the nic.")
+    if _have_attachment(db, nic, model.NetworkAttachment.channel == channel):
+        raise BlockedError("The channel is already in use on the nic.")
+
+    if not allocator.is_legal_channel_for(db, channel, network.network_id):
+        raise BadArgumentError("Channel %r, is not legal for this network." %
+                               channel)
+
```

Comment:
Started working on this, and found there was actually a test that checks for the behavior you're pointing out. Talked to @gsilvis, I think the reasoning at the time was that you should never actually have nics in the DB that aren't connected to the switch, and that this was an administrator error. The driver writes a warning to the log and just noops. I'm not sure I still think this is the right behaviour, but let's not add unrelated api changes to this patch.